File: src/Servers/ServerBrowser/src/Entity/Structure/Misc/Encryption.cs
using System.Collections.Generic;
using UniSpyServer.UniSpyLib.Encryption;
using UniSpyServer.ServerBrowser.Entity.Structure.Misc;
namespace UniSpyServer.ServerBrowser.Entity.Structure.Misc
{
    public class EncryptionParameters
    {
        public byte[] Register = new byte[256];
        public byte Index0;
        public byte Index1;
        public byte Index2;
        public byte Index3;
        public byte Index4;
    }

    public class SBEncryption
    {
        private EncryptionParameters _encParams;
        private byte[] _clientChallenge = new byte[8];
        private byte[] _serverChallenge;
        private byte[] _secretKey;
        public SBEncryption(string secretKey, string clientChallenge, EncryptionParameters encParam)
        {
            // reference private filed to encParams
            _encParams = encParam;
            _clientChallenge = UniSpyEncoding.GetBytes(clientChallenge);
            _serverChallenge = UniSpyEncoding.GetBytes(Constants.ServerChallenge);
            _secretKey = UniSpyEncoding.GetBytes(secretKey);
            InitEncryptionAlgorithm();
        }

        public SBEncryption(EncryptionParameters encParams)
        {
            _encParams = encParams;
        }

        /// <summary>
        /// Initialize the challenge which generates the permutation table
        /// </summary>
        /// <param name="secKey">game secret key</param>
        /// <param name="serverChallenge">the random number challenge that server or client generate</param>
        public void InitEncryptionAlgorithm()
        {
            if (_clientChallenge.Length != 8)
            {
                //we have error client challenge
                throw new System.ArgumentException("Client challenge length not valid!");
            }

            for (int i = 0; i < _serverChallenge.Length; i++)
            {
                int tempIndex0 = i * _secretKey[i % _secretKey.Length] % 8;
                int tempIndex1 = i % 8;
                byte bitwiseResult = (byte)(_clientChallenge[tempIndex1] ^ _serverChallenge[i]);
                _clientChallenge[tempIndex0] ^= (byte)(bitwiseResult & 0xFF);
            }

            InitEncryptionParameters();
        }

        private void InitEncryptionParameters()
        {
            uint i;
            byte toSwap, swapTemp, randomSum;
            byte keyPosition;

            if (_clientChallenge.Length < 1)
            {
                NonChallengeMappingInit();
                return;
            }

            //generate ascii table
            for (i = 0; i < 256; i++)
            {
                _encParams.Register[i] = (byte)i;
            }

            toSwap = 0;
            keyPosition = 0;
            randomSum = 0;
            for (i = 255; i > 0; i--)
            {
                toSwap = IndexPositionGeneration(i, ref randomSum, ref keyPosition);
                swapTemp = _encParams.Register[i];
                _encParams.Register[i] = _encParams.Register[toSwap];
                _encParams.Register[toSwap] = swapTemp;
            }

            // Initialize the indices and data dependencies.
            // Indices are set to different values instead of all 0
            // to reduce what is known about the state of the state.cards
            // when the first byte is emitted.

            _encParams.Index0 = _encParams.Register[1];
            _encParams.Index1 = _encParams.Register[3];
            _encParams.Index2 = _encParams.Register[5];
            _encParams.Index3 = _encParams.Register[7];
            _encParams.Index4 = _encParams.Register[randomSum];
        }

        public byte[] Encrypt(string plainStr)
        {
            byte[] plainText = UniSpyEncoding.GetBytes(plainStr);
            return Encrypt(plainText);
        }

        public byte[] Encrypt(byte[] plainText)
        {
            int i;
            byte[] cipherText = new byte[plainText.Length];
            for (i = 0; i < plainText.Length; i++)
            {
                cipherText[i] = ByteShift(plainText[i]);
            }
            return cipherText;
        }

        private void NonChallengeMappingInit()
        {
            // Initialize the indices and data dependencies.
            byte i, j;
            _encParams.Index0 = 1;
            _encParams.Index1 = 3;
            _encParams.Index2 = 5;
            _encParams.Index3 = 7;
            _encParams.Index4 = 11;

            // Start with state.cards all in inverse order.
            for (i = 0, j = 255; i <= 255; i++, j--)
            {
                _encParams.Register[i] = j;
            }
        }

        private byte ByteShift(byte b)
        {
            byte swapTempStorage;
            _encParams.Index1 = (byte)(_encParams.Index1 + _encParams.Register[_encParams.Index0++]);
            swapTempStorage = _encParams.Register[_encParams.Index4];
            _encParams.Register[_encParams.Index4] = _encParams.Register[_encParams.Index1];
            _encParams.Register[_encParams.Index1] = _encParams.Register[_encParams.Index3];
            _encParams.Register[_encParams.Index3] = _encParams.Register[_encParams.Index0];
            _encParams.Register[_encParams.Index0] = swapTempStorage;
            _encParams.Index2 = (byte)(_encParams.Index2 + _encParams.Register[swapTempStorage]);


            _encParams.Index4 = (byte)(b ^ _encParams.Register[(_encParams.Register[_encParams.Index2] +
                _encParams.Register[_encParams.Index0]) & 0xFF] ^
               _encParams.Register[_encParams.Register[(_encParams.Register[_encParams.Index3] +
               _encParams.Register[_encParams.Index4] +
               _encParams.Register[_encParams.Index1]) & 0xFF]]);
            _encParams.Index3 = b;

            return _encParams.Index4;
        }


        /// <summary>
        /// Generate permutation index position
        /// </summary>
        /// <param name="limit"></param>
        /// <param name="challenge"></param>
        /// <param name="randomSum"></param>
        /// <param name="keyPosition"></param>
        /// <returns></returns>
        private byte IndexPositionGeneration(uint limit, ref byte randomSum, ref byte keyPosition)
        {
            byte swapIndex, retryLimiter, bitMask;

            if (limit == 0)
            {
                return 0;
            }

            retryLimiter = 0;
            bitMask = 1;

            while (bitMask < limit)
            {
                bitMask = (byte)((bitMask << 1) + 1);
            }

            do
            {
                randomSum = (byte)(_encParams.Register[randomSum] + _clientChallenge[keyPosition++]);

                if (keyPosition >= _clientChallenge.Length)
                {
                    keyPosition = 0;
                    randomSum = (byte)(randomSum + _clientChallenge.Length);
                }
                swapIndex = (byte)(bitMask & randomSum);
                if (++retryLimiter > 11)
                {
                    swapIndex %= (byte)limit;
                }
            }
            while (swapIndex > limit);

            return swapIndex;
        }
        public static byte[] BuildCryptHeader()
        {
            // cryptHeader have 14 bytes, when we encrypt data we need skip the first 14 bytes
            var cryptHeader = new List<byte>();
            cryptHeader.Add(2 ^ 0xEC);
            cryptHeader.AddRange(new byte[] { 0, 0 });
            cryptHeader.Add((byte)(Constants.ServerChallenge.Length ^ 0xEA));
            cryptHeader.AddRange(UniSpyEncoding.GetBytes(Constants.ServerChallenge));
            return cryptHeader.ToArray();
        }
    }
}

File: src/frontends/gamespy/protocols/server_browser/v2/aggregations/encryption.py
from frontends.gamespy.library.abstractions.enctypt_base import Byte, EncryptBase

SERVER_CHALLENGE = "0000000000"


class EncryptionParameters:
    register: list[Byte]
    index_0: Byte
    index_1: Byte
    index_2: Byte
    index_3: Byte
    index_4: Byte

    def __init__(self):
        self.register = [Byte(i) for i in range(256)]
        self.index_0 = Byte(0)
        self.index_1 = Byte(0)
        self.index_2 = Byte(0)
        self.index_3 = Byte(0)
        self.index_4 = Byte(0)


def clamp(value: int) -> int:
    return value % 256


class EnctypeX(EncryptBase):
    _enc_params: EncryptionParameters
    _clientChallenge: list[Byte]
    _serverChallenge: list[Byte]
    _secretKey: list[Byte]

    def __init__(self, secretKey: str, clientChallenge: str):
        assert isinstance(secretKey, str)
        assert isinstance(clientChallenge, str)
        self._enc_params = EncryptionParameters()
        self._clientChallenge = Byte.from_bytes(
            clientChallenge.encode("ascii"))
        self._serverChallenge = Byte.from_bytes(
            SERVER_CHALLENGE.encode("ascii"))
        self._secretKey = Byte.from_bytes(secretKey.encode("ascii"))
        self.init_encryption_algorithm()

    def init_encryption_algorithm(self):
        if len(self._clientChallenge) != 8:
            raise ValueError("Client challenge length not valid!")

        for i in range(len(self._serverChallenge)):
            temp_index_0 = Byte(
                i) * self._secretKey[i % len(self._secretKey)] % Byte(8)
            temp_index1 = Byte(i % 8)
            bitwise_result = self._clientChallenge[temp_index1.value] ^ self._serverChallenge[i]
            self._clientChallenge[temp_index_0.value] ^= bitwise_result

        self.init_encryption_parameters()

    def init_encryption_parameters(self):
        if len(self._clientChallenge) < 1:
            self.non_challenge_mapping_init()
            return

        to_swap = Byte(0)
        key_position = Byte(0)
        random_sum = Byte(0)
        for i in range(255, 0, -1):
            random_sum, key_position, to_swap = self.index_position_generation(
                Byte(i), random_sum, key_position)
            swap_temp = self._enc_params.register[i]
            self._enc_params.register[i] = self._enc_params.register[to_swap.value]
            self._enc_params.register[to_swap.value] = swap_temp

        self._enc_params.index_0 = self._enc_params.register[1]
        self._enc_params.index_1 = self._enc_params.register[3]
        self._enc_params.index_2 = self._enc_params.register[5]
        self._enc_params.index_3 = self._enc_params.register[7]
        self._enc_params.index_4 = self._enc_params.register[random_sum.value]

    def non_challenge_mapping_init(self):
        self._enc_params.index_0 = Byte(1)
        self._enc_params.index_1 = Byte(3)
        self._enc_params.index_2 = Byte(5)
        self._enc_params.index_3 = Byte(7)
        self._enc_params.index_4 = Byte(11)
        self._enc_params.register = [Byte(i) for i in range(256, 0, -1)]

    def byte_shift(self, b: Byte) -> Byte:
        self._enc_params.index_1 += self._enc_params.register[self._enc_params.index_0.value]
        self._enc_params.index_0 += Byte(1)
        swap_temp_storage = self._enc_params.register[self._enc_params.index_4.value]
        self._enc_params.register[self._enc_params.index_4.value] = self._enc_params.register[
            self._enc_params.index_1.value
        ]
        self._enc_params.register[self._enc_params.index_1.value] = self._enc_params.register[
            self._enc_params.index_3.value
        ]
        self._enc_params.register[self._enc_params.index_3.value] = self._enc_params.register[
            self._enc_params.index_0.value
        ]
        self._enc_params.register[self._enc_params.index_0.value] = swap_temp_storage
        self._enc_params.index_2 += self._enc_params.register[swap_temp_storage.value]

        part1 = (
            self._enc_params.register[self._enc_params.index_2.value]
            + self._enc_params.register[self._enc_params.index_0.value]
        ).value % 256
        part2 = (self._enc_params.register[self._enc_params.index_3.value].value
                 + self._enc_params.register[self._enc_params.index_4.value].value
                 + self._enc_params.register[self._enc_params.index_1.value].value) % 256
        part3 = self._enc_params.register[part2].value
        self._enc_params.index_4 = Byte(
            b.value ^ self._enc_params.register[part1].value
            ^ self._enc_params.register[part3].value
        )
        self._enc_params.index_3 = b

        return self._enc_params.index_4

    def index_position_generation(self, limit: Byte, random_sum: Byte, key_position: Byte) -> tuple[Byte, Byte, Byte]:
        swap_index, retry_limiter, bit_mask = Byte(0), Byte(0), Byte(1)
        if limit == Byte(0):
            return random_sum, key_position, Byte(0)

        while bit_mask < limit:
            bit_mask = (bit_mask << Byte(1)) + Byte(1)

        while True:
            random_sum = (
                self._enc_params.register[random_sum.value] +
                self._clientChallenge[key_position.value]
            )
            key_position += Byte(1)

            if key_position >= Byte(len(self._clientChallenge)):
                key_position = Byte(0)
                random_sum += Byte(len(self._clientChallenge))

            swap_index = bit_mask & random_sum

            if retry_limiter > Byte(11):
                swap_index %= limit
            retry_limiter += Byte(1)

            if swap_index <= limit:
                break

        return random_sum, key_position, swap_index

    def encrypt(self, plain_text: bytes):
        # skip first 14 bytes
        head_buffer = plain_text[:14]
        body_buffer = plain_text[14:]
        cipher_body = bytearray(body_buffer)
        for i in range(len(body_buffer)):
            c = self.byte_shift(Byte(body_buffer[i]))
            cipher_body[i] = c.value
        cipher_body = bytes(cipher_body)
        cipher = head_buffer+cipher_body
        return cipher
